### PR TITLE
Add replace header function

### PR DESCRIPTION
**Pull request by yuf3n9**

### CODE REVIEW

File: gmime/envelope.go [hunk 1]
```diff
@@ -69,6 +69,49 @@ func (m *Envelope) Headers() textproto.MIMEHeader {
...
+		if name != key {
+			continue
+		}
+		value := C.GoString(C.g_mime_header_get_value(header))
+		if value != originalValue {
+			continue
+		}
+		format := C.g_mime_format_options_get_default()
+		cCharset := C.CString("UTF-8")
+		C.g_mime_header_set_value(header, format, C.CString(replaceValue), cCharset)
```
Comment (Dharmik-Patel) [inline gmime/envelope.go]: does set value return any errors?

File: gmime/envelope.go [hunk 2]
```diff
@@ -69,6 +69,49 @@ func (m *Envelope) Headers() textproto.MIMEHeader {
 	return goHeaders
 }
 
+// ReplaceHeader replaces the header with matching key & originalValue with replaceValue
+func (m *Envelope) ReplaceHeader(key, originalValue, replaceValue string) error {
+	// TODO: this is not super efficient, but easier to read, may be optimize this?
```
Comment (Dharmik-Patel) [inline gmime/envelope.go]: remove this

File: gmime/envelope.go (hunk 1 above)
Comment (yuf3n9) [reply gmime/envelope.go]: nope
https://github.com/GNOME/gmime/blob/3b71f3ff9e4d8882b685ba8071945f80833b62c5/gmime/gmime-header.c#L252

File: gmime/gmime_test.go [hunk 3]
```diff
@@ -183,6 +186,74 @@ func TestRemoveAll(t *testing.T) {
 	assert.Equal(t, "Kien Pham <kien@sendgrid.com>", msg.Header("To"))
 }
 
+func TestReplaceHeader(t *testing.T) {
```
Comment (Dharmik-Patel) [inline gmime/gmime_test.go:189]: can you add test description for each of the tests or do table driven tests

File: gmime/envelope.go [hunk 4]
```diff
@@ -69,6 +69,29 @@ func (m *Envelope) Headers() textproto.MIMEHeader {
 	return goHeaders
 }
 
+// ReplaceHeader replaces the header with matching key & originalValue with replaceValue
+func (m *Envelope) ReplaceHeader(key, originalValue, replaceValue string) error {
```
Comment (Dharmik-Patel) [inline gmime/envelope.go:73]: can you run this in a forloop and check for memory leaks?

Comment (yuf3n9) [reply gmime/envelope.go:73]: made a fix. memory usage is constant (around 6.5MB) in forloop now

File: gmime/gmime_test.go [hunk 5]
```diff
@@ -183,6 +186,79 @@ func TestRemoveAll(t *testing.T) {
...
+}
+
+type headerData struct {
+	name  string
+	value string
+}
+
+func headersSlice(mimeBytes []byte) ([]headerData, error) {
+	var headers []headerData
+	b := bytes.NewReader(mimeBytes)
```
Comment (rhansen2) [inline gmime/gmime_test.go:238]: Will this work with headers wrapped over multiple lines?

File: gmime/envelope.go [hunk 6]
```diff
@@ -69,6 +69,32 @@ func (m *Envelope) Headers() textproto.MIMEHeader {
...
+	headers := C.g_mime_object_get_header_list(m.asGMimeObject())
+	count := C.g_mime_header_list_get_count(headers)
+	var i C.int
+	for i = 0; i < count; i++ {
+		header := C.g_mime_header_list_get_header_at(headers, i)
+		name := C.GoString(C.g_mime_header_get_name(header))
```
Comment (rhansen2) [inline gmime/envelope.go:79]: Do we need to free any of these strings?

Comment (yuf3n9) [reply gmime/envelope.go:79]: i think only CString needs to be freed

Comment (rhansen2) [reply gmime/envelope.go:79]: Doesn't this line create a C string then convert it to a Go string? It isn't alway clear to me what needs to be freed in gmime and what it takes care of automatically.

Comment (yuf3n9) [reply gmime/envelope.go:79]: ahh. yeah, these are part of headerList so gmime takes care of it.
https://developer.gnome.org/gmime/stable/GMimeObject.html#g-mime-object-get-header-list
> the GMimeHeaderList for object . Do not free this pointer when you are done with it.

Comment (rhansen2) [reply gmime/envelope.go:79]: What about: g_mime_header_get_name?

Comment (yuf3n9) [reply gmime/envelope.go:79]: it returns `header->name`, so I assume as gmime frees headerList, name is also freed?

Comment (rhansen2) [reply gmime/envelope.go:79]: in that case probably.

**Pull request merged**